beam-core/Database/Beam/Query/Combinators.hs

```diff
@@ -335,6 +337,15 @@ limit_ :: forall s a be db
 limit_ limit' (Q q) =
   Q (liftF (QLimit limit' q (rewriteThread (Proxy @s))))
 
+-- | Conditionally limit the number of results returned by a query.
+limitMaybe_ :: forall s a be db
+            . ( Projectible be a
+              , ThreadRewritable (QNested s) a )
+            => Maybe Integer -> Q be db (QNested s) a -> Q be db s (WithRewrittenThread (QNested s) s a)
+limitMaybe_ (Just limit') (Q q) =
+  Q (liftF (QLimit limit' q (rewriteThread (Proxy @s))))
+limitMaybe_ Nothing (Q q) = Q (unsafeCoerce q)
```

Comment:
I think you should be able to use `rewriteThread`

Reply:
Sorry, could you give a hint on how that would look? I haven't grokked this stuff yet...

Reply:
@tathougies I'm going through old PRs in and was blocked on this question -- could you clarify?

Reply:
Yeah, so the type doesn't match I believe because `(QNested s)` doesn`t unify with `WithRewrittenThread (QNested s) s`. The latter unwraps one level of `QNested`, which is what the rewriteThread function is fro. I believe if you do.

`rewriteThread (Proxy @s) q` instead of the `unsafeCoerce q` , that should work.

Reply:
I tried some stuff like that and wasn't able to get it to work:

```shellSession
Building library for beam-core-0.9.2.1..
[28 of 32] Compiling Database.Beam.Query.Combinators

/home/tom/tools/beam/beam-core/Database/Beam/Query/Combinators.hs:347:32: error:
    • Couldn't match type: WithRewrittenThread
                             s0 s (QM be db (QNested s) a)
                     with: Control.Monad.Free.Church.F
                             (QF be db s) (WithRewrittenThread (QNested s) s a)
      Expected: QM be db s (WithRewrittenThread (QNested s) s a)
        Actual: WithRewrittenThread s0 s (QM be db (QNested s) a)
      The type variable ‘s0’ is ambiguous
    • In the first argument of ‘Q’, namely
        ‘(rewriteThread (Proxy @s) q)’
      In the expression: Q (rewriteThread (Proxy @s) q)
      In an equation for ‘limitMaybe_’:
          limitMaybe_ Nothing (Q q) = Q (rewriteThread (Proxy @s) q)
    • Relevant bindings include
        q :: QM be db (QNested s) a
          (bound at Database/Beam/Query/Combinators.hs:347:24)
        limitMaybe_ :: Maybe Integer
                       -> Q be db (QNested s) a
                       -> Q be db s (WithRewrittenThread (QNested s) s a)
          (bound at Database/Beam/Query/Combinators.hs:345:1)
    |
347 | limitMaybe_ Nothing (Q q) = Q (rewriteThread (Proxy @s) q)
    |                                ^^^^^^^^^^^^^^^^^^^^^^^^^^
```

Reply:
@tathougies pinging again, would love to get this merged

Reply:
@tathougies another ping, please help! or maybe we could just merge as-is?

Reply:
@tathougies here is the ping for 2024 :)